Mathlib/RingTheory/TwoSidedIdeal.lean

```diff
@@ -0,0 +1,227 @@
+
+lemma add_mem {x y} (hx : x ∈ I) (hy : y ∈ I) : x + y ∈ I := by
+  simpa using I.add hx hy
+
+lemma neg_mem {x} (hx : x ∈ I) : -x ∈ I := by
+  simpa using I.neg hx
+
+lemma sub_mem {x y} (hx : x ∈ I) (hy : y ∈ I) : x - y ∈ I := by
+  simpa using I.sub hx hy
+
+lemma mul_mem_left (x y) (hy : y ∈ I) : x * y ∈ I := by
+  simpa using I.mul (I.refl x) hy
+
+lemma mul_mem_right (x y) (hx : x ∈ I) : x * y ∈ I := by
+  simpa using I.mul hx (I.refl y)
+
+lemma nsmul_mem {x} (n : ℕ) (hx : x ∈ I) : n • x ∈ I := by
+  simpa using I.nsmul _ hx
+
+lemma zsmul_mem {x} (n : ℤ) (hx : x ∈ I) : n • x ∈ I := by
+  simpa using I.zsmul _ hx
+
+instance : AddSubgroupClass (RingCon R) R where
+  zero_mem := zero_mem
+  add_mem := @add_mem _ _
+  neg_mem := @neg_mem _ _
+
+instance : Add I where add x y := ⟨x.1 + y.1, I.add_mem x.2 y.2⟩
+
+instance : Zero I where zero := ⟨0, I.zero_mem⟩
+
+instance : SMul ℕ I where
+  smul n x := ⟨n • x.1, I.nsmul_mem n x.2⟩
+
+instance : Neg I where neg x := ⟨-x.1, I.neg_mem x.2⟩
+
+instance : Sub I where sub x y := ⟨x.1 - y.1, I.sub_mem x.2 y.2⟩
+
+instance : SMul ℤ I where
+  smul n x := ⟨n • x.1, I.zsmul_mem n x.2⟩
+
+instance : AddCommGroup I :=
+  Function.Injective.addCommGroup _ Subtype.coe_injective
+    rfl (fun _ _ ↦ rfl) (fun _ ↦ rfl) (fun _ _ ↦ rfl) (fun _ _ ↦ rfl) (fun _ _ ↦ rfl)
+
+/-- The coercion into the ring as a `AddMonoidHom` -/
+@[simp]
+def coeAddMonoidHom : I →+ R where
+  toFun := (↑)
+  map_zero' := rfl
+  map_add' _ _ := rfl
+
+end NonUnitalNonAssocRing
+
+section ring
+
+variable {R : Type*} [Ring R] (I : RingCon R)
+
+instance : SMul R I where smul r x := ⟨r • x.1, I.mul_mem_left _ _ x.2⟩
+
+instance : SMul Rᵐᵒᵖ I where smul r x := ⟨r • x.1, I.mul_mem_right _ _ x.2⟩
+
+instance : Module R I :=
+  Function.Injective.module _ (coeAddMonoidHom I) Subtype.coe_injective fun _ _ => rfl
+
+instance : Module Rᵐᵒᵖ I :=
+  Function.Injective.module _ (coeAddMonoidHom I) Subtype.coe_injective fun _ _ => rfl
+
+instance : SMulCommClass R Rᵐᵒᵖ I where
+  smul_comm r s x := Subtype.ext <| smul_comm r s x.1
+
+/--
+For any `I : RingCon R`, when we view it as an ideal, `I.subtype` is the injective `R`-linear map
+`I → R`.
+-/
+@[simps]
+def subtype : I →ₗ[R] R where
+  toFun x := x.1
+  map_add' _ _ := rfl
+  map_smul' _ _ := rfl
+
+/--
+For any `RingCon R`, when we view it as an ideal in `Rᵒᵖ`, `subtype` is the injective `Rᵐᵒᵖ`-linear
+map `I → Rᵐᵒᵖ`.
+-/
+@[simps]
+def subtypeMop : I →ₗ[Rᵐᵒᵖ] Rᵐᵒᵖ where
+  toFun x := MulOpposite.op x.1
+  map_add' _ _ := rfl
+  map_smul' _ _ := rfl
+
+end ring
+
+section operations
+
+variable {R : Type*} [Ring R] (I : RingCon R)
+
+/--
+Every two-sided ideal is also a left ideal.
+-/
+def toIdeal : Ideal R where
+  carrier := I
+  add_mem' := I.add_mem
+  zero_mem' := I.zero_mem
+  smul_mem' := I.mul_mem_left
+
+lemma mem_toIdeal {x} : x ∈ I.toIdeal ↔ x ∈ I := Iff.rfl
+
+/--
+Every two-sided ideal is also a right ideal.
+-/
+def toIdealMop : Ideal Rᵐᵒᵖ := LinearMap.range I.subtypeMop
```

Comment:
This gives bad defeqs. Maybe first show `RingCon Rᵐᵒᵖ = RinCon R` and then use `toIdeal`.

Reply:
I don’t think I understand. Isn’t type equality even more evil.

Reply:
Sorry I meant equivs

Reply:
I have made a change, but I still can't get
```
lemma mem_toIdealMop {x} : x ∈ I.toIdealMop ↔ x.unop ∈ I := by
```
to be true by Iff.rfl